### DISCUSSION TITLE
Articial Viscosity / Diffusion based stabilisation

### DISCUSSION THREAD

**Discussion by j-bowhay** [idaholab/moose#25237] [Q&A General] [answered]
Hi All

I am modelling a problem in MOOSE that involves solving a hyperbolic PDE. I have been having good success at stabilizing the solution by adding an artificial diffusion / viscosity term $\nu \Delta u$. Currently, I am applying this indiscriminately to the whole domain but I want to make this method more refined. Specifically, I want the viscosity term to be a function of the residual ie:
<img width="214" alt="Screenshot 2023-08-17 at 16 56 59" src="https://github.com/idaholab/moose/assets/60778417/e073946d-b781-4782-ade7-11adcd53949d">
where the residual for the 1d transport eq would be $r(x)=u_t(x)+u_x(x)$. I am unsure how best to implement this in MOOSE, should I just calculate $r(x)$ directly or can I use the tagging system?

Many Thanks
Jake

**Comment by GiudGiud** [comment]
Hello

I would calculate r directly. 

We don’t compute the residual in two steps so on paper we would not have access to the residual when computing a term of the residual. 

However,
You could either:
- try to use a projection of the previous residual by tagging it then copying it at the end of a time step in an aux variable. That might not be great
- Using the show_exeucution_order Boolean option in the Debug block you could make sure this kernel is calculated last. This is iffy but could work. There s no réal guarantee on the order of execution of kernels
- 
Guillaume

**Reply by j-bowhay** [reply]
Thank @GiudGiud, I will proceed with calculating r directly. Is there a way to access old values of $\nabla u$ like there is for $\dot{u}$ (`_var.uDotOld()`)?

Edit: believe I can use `_var.gradSlnOld()`

**Reply by GiudGiud** [reply]
you can use this too:
https://mooseframework.inl.gov/docs/doxygen/moose/classCoupleable.html#aafe7b956f1dbe65960e5b6620f6a1236

It relies on the coupleable interface, which requires you to add your variable in the coupled variable parameters and also initialize a member variable of the class with coupledGradientOld

**Comment by lindsayad** [comment, marked as answer]
In the Navier Stokes SUPG implementation we compute strong form residuals in materials and reuse those in 1) terms that are not integrated by parts and 2) the kernel that adds the SUPG stabilization. Perhaps you could consider a similar design
